cmd/src/debug_comp.go

```diff
@@ -0,0 +1,270 @@
+		if err != nil {
+			return cmderrors.ExitCode(1, nil)
+		}
+		return nil
+	}
+
+	debugCommands = append(debugCommands, &command{
+		flagSet: flagSet,
+		handler: handler,
+		usageFunc: func() {
+			fmt.Println(usage)
+		},
+	})
+}
+
+/*
+Docker functions
+TODO: handle for single container/server instance
+*/
+
+func archiveDocker(ctx context.Context, zw *zip.Writer, verbose, configs bool, baseDir string) error {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	containers, err := getContainers(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to get docker containers: %w", err)
+	}
+
+	if verbose {
+		log.Printf("getting docker data for %d containers...\n", len(containers))
+	}
+
+	// setup channel for slice of archive function outputs
+	ch := make(chan *archiveFile)
+	wg := sync.WaitGroup{}
+	semaphore := semaphore.NewWeighted(8)
+
+	// start goroutine to run docker ps -o wide
+	wg.Add(1)
+	go func() {
+		if err := semaphore.Acquire(ctx, 1); err != nil {
+			return
+		}
+		defer semaphore.Release(1)
+		defer wg.Done()
+		ch <- getPs(ctx, baseDir)
+	}()
+
+	// start goroutine to run docker container stats --no-stream
+	wg.Add(1)
+	go func() {
+		if err := semaphore.Acquire(ctx, 1); err != nil {
+			return
+		}
+		defer semaphore.Release(1)
+		defer wg.Done()
+		ch <- getStats(ctx, baseDir)
+	}()
+
+	// start goroutine to run docker container logs <container>
+	for _, container := range containers {
+		wg.Add(1)
+		go func(container string) {
+			if err := semaphore.Acquire(ctx, 1); err != nil {
+				return
+			}
+			defer semaphore.Release(1)
+			defer wg.Done()
+			ch <- getContainerLog(ctx, container, baseDir)
+		}(container)
+	}
+
+	// start goroutine to run docker container inspect <container>
+	for _, container := range containers {
+		wg.Add(1)
+		go func(container string) {
+			if err := semaphore.Acquire(ctx, 1); err != nil {
+				return
+			}
+			defer semaphore.Release(1)
+			defer wg.Done()
+			ch <- getInspect(ctx, container, baseDir)
+		}(container)
+	}
+
+	// start goroutine to get configs
+	if configs {
+		wg.Add(1)
+		go func() {
+			if err := semaphore.Acquire(ctx, 1); err != nil {
+				return
+			}
+			defer semaphore.Release(1)
+			defer wg.Done()
+			ch <- getExternalServicesConfig(ctx, baseDir)
+		}()
+
+		wg.Add(1)
+		go func() {
+			if err := semaphore.Acquire(ctx, 1); err != nil {
+				return
+			}
+			defer semaphore.Release(1)
+			defer wg.Done()
+			ch <- getSiteConfig(ctx, baseDir)
+		}()
+	}
+
+	// close channel when wait group goroutines have completed
+	go func() {
+		wg.Wait()
+		close(ch)
+	}()
+
+	for f := range ch {
+		if f.err != nil {
+			return fmt.Errorf("aborting due to error on %s: %v\noutput: %s", f.name, f.err, f.data)
+		}
+
+		if verbose {
+			log.Printf("archiving file %q with %d bytes", f.name, len(f.data))
+		}
+
+		zf, err := zw.Create(f.name)
+		if err != nil {
+			return fmt.Errorf("failed to create %s: %w", f.name, err)
+		}
+
+		_, err = zf.Write(f.data)
```

Comment:
Stylistic opinion (not everyone at sg agrees): if all you're producing is an `err` value, I prefer to write the check inline:

```go
if _, err := failer(); err != nil { ... }
```

rather than separating them:

```go
_, err = failer()
if err != nil { ... }
```

Since `err` is such a common name, the `if` creates a new scope that doesn't influence the rest of the function. This can save a lot of `=` <-> `:=` transformations when changing unrelated code.